test/wasm2c/tail-calls.txt

```diff
@@ -0,0 +1,953 @@
+    va_start(args, result_count);
+    if (true && va_arg(args, wasm_rt_type_t) == WASM_RT_I32 && va_arg(args, wasm_rt_type_t) == WASM_RT_F64 && va_arg(args, wasm_rt_type_t) == WASM_RT_I32 && va_arg(args, wasm_rt_type_t) == WASM_RT_F64) {
+      va_end(args);
+      return w2c_test_t2;
+    }
+    va_end(args);
+  }
+  
+  return NULL;
+}
+
+void w2c_test_tailcaller_0(w2c_test* instance) {
+  FUNC_PROLOGUE;
+  u32 var_i0, var_i2;
+  f32 var_f1;
+  var_i0 = 1u;
+  var_f1 = 2;
+  var_i2 = 0u;
+  static_assert(sizeof(struct wasm_multi_if) <= 1024);
+  CHECK_CALL_INDIRECT(instance->w2c_tab, w2c_test_i32_f32, var_i2);
+  if (!instance->w2c_tab.data[var_i2].func_tailcallee.fn) {
+    CALL_INDIRECT(instance->w2c_tab, void (*)(void*, u32, f32), w2c_test_i32_f32, var_i2, instance->w2c_tab.data[var_i2].module_instance, var_i0, var_f1);
+  } else {
+    void *instance_ptr_storage;
+    void **instance_ptr = &instance_ptr_storage;
+    char tail_call_stack[1024];
+    wasm_rt_tailcallee_t next_storage;
+    wasm_rt_tailcallee_t *next = &next_storage;
+    {
+      struct wasm_multi_if *tmp = (struct wasm_multi_if *)tail_call_stack;
+      tmp->i0 = var_i0;
+      tmp->f1 = var_f1;
+    }
+    next->fn = instance->w2c_tab.data[var_i2].func_tailcallee.fn;
+    *instance_ptr = instance->w2c_tab.data[var_i2].module_instance;
+    while (next->fn) { next->fn(instance_ptr, tail_call_stack, next); }
+  }
+  goto var_Bfunc;
+  var_Bfunc:;
+  FUNC_EPILOGUE;
+}
+
+void wasm_tailcall_w2c_test_tailcaller_0(void **instance_ptr, void *tail_call_stack, wasm_rt_tailcallee_t *next) {
+  w2c_test* instance = *instance_ptr;
+  u32 var_i0, var_i2;
+  f32 var_f1;
+  var_i0 = 1u;
+  var_f1 = 2;
+  var_i2 = 0u;
+  static_assert(sizeof(struct wasm_multi_if) <= 1024);
+  CHECK_CALL_INDIRECT(instance->w2c_tab, w2c_test_i32_f32, var_i2);
+  if (!instance->w2c_tab.data[var_i2].func_tailcallee.fn) {
+    CALL_INDIRECT(instance->w2c_tab, void (*)(void*, u32, f32), w2c_test_i32_f32, var_i2, instance->w2c_tab.data[var_i2].module_instance, var_i0, var_f1);
+    next->fn = NULL;
+  } else {
+    {
+      struct wasm_multi_if *tmp = (struct wasm_multi_if *)tail_call_stack;
+      tmp->i0 = var_i0;
+      tmp->f1 = var_f1;
+    }
+    next->fn = instance->w2c_tab.data[var_i2].func_tailcallee.fn;
+    *instance_ptr = instance->w2c_tab.data[var_i2].module_instance;
+  }
+  return;
+  var_Bfunc:;
+  next->fn = NULL;
+}
+
+struct wasm_multi_id w2c_test_infiniteloop(w2c_test* instance, u32 var_p0, f64 var_p1) {
+  FUNC_PROLOGUE;
+  u32 var_i0;
+  f64 var_d1;
+  var_i0 = var_p0;
+  var_d1 = var_p1;
+  static_assert(sizeof(struct wasm_multi_id) <= 1024);
+  {
+    void *instance_ptr_storage;
+    void **instance_ptr = &instance_ptr_storage;
+    char tail_call_stack[1024];
+    wasm_rt_tailcallee_t next_storage;
+    wasm_rt_tailcallee_t *next = &next_storage;
+    {
+      struct wasm_multi_id *tmp = (struct wasm_multi_id *)tail_call_stack;
+      tmp->i0 = var_i0;
```

Comment:
Hmm, I don't _think_ this is undefined behavior -- every object's value is being accessed as the same type as when it was stored-to or modified. And if not, I don't think an explicit `memcpy` would fix it -- it would still be UB afaik.

(Footnote: That's assuming the module is valid, which the command-line wasm2c tool enforces, but CWriter just assumes as a precondition. I think if we're worried about invalid modules, CWriter probably would need bolstering in a lot of places including this one. :-/)

Reply:
Hmm... We may need to go through this a bit more before I am fully convinced. Here is my reasoning below for why this could be UB. I want to make sure this doesn't apply before we commit. Let me know your thoughts.

Of the top of my head, I believe the high level reason I think this maybe an issue, is that it is in-effect a use of type punning, which I believe in general is UB, even with the property "every object's value is being accessed as the same type as when it was stored-to or modified". This property is sufficient for the safe use of union types, but without union types, I think this may not be enough. 

I believe the heart of this discrepancy is that C can assume that pointers of different types do not alias, so any code of the form 

```c
struct wasm_multi_id* a = (struct wasm_multi_id*) tail_call_stack;
...
struct wasm_multi_ii* b = (struct wasm_multi_ii*) tail_call_stack;
```

is compiled assuming `a` and `b` don't alias. 

Fwiw, i did run this through uban and it didn't complain, so maybe its ok? However, this pattern makes me nervous. I think with local vars + memcpy, there is no possibility of the code being UB as I believe the C standard says this is always safe much more directly. In past projects, I have seen that compiler is definitely smart enough to optimize these extra memcpy's and local vars away, so its unlikely to be a performance hit. Given this, I think this maybe the best course of action here.

In case its useful, I wroteup a small godbolt demo for how to save the contents of different structs in shared char buffer (both how we do this currently and how to do it with local vars without UB)

https://godbolt.org/z/79jWbcWPb

Let me know your thoughts!

Reply:
(Also found a couple of links when searching
https://stackoverflow.com/questions/63422076/why-is-type-punning-considered-ub
https://adriann.github.io/undefined_behavior.html)

Reply:
Ok, changed it to use memcpy everywhere.